programs/dibio.c

```diff
@@ -260,18 +324,31 @@ int DiB_trainFromFiles(const char* dictFileName, unsigned maxDictSize,
                        ZDICT_legacy_params_t* params, ZDICT_cover_params_t* coverParams,
                        ZDICT_fastCover_params_t* fastCoverParams, int optimize)
 {
-    unsigned const displayLevel = params ? params->zParams.notificationLevel :
-                        coverParams ? coverParams->zParams.notificationLevel :
-                        fastCoverParams ? fastCoverParams->zParams.notificationLevel :
-                        0;   /* should never happen */
+    fileStats fs;
     void* const dictBuffer = malloc(maxDictSize);
-    fileStats const fs = DiB_fileStats(fileNamesTable, nbFiles, chunkSize, displayLevel);
+
+    int const displayLevel = params ? params->zParams.notificationLevel :
+        coverParams ? coverParams->zParams.notificationLevel :
+        fastCoverParams ? fastCoverParams->zParams.notificationLevel : 0;
```

Comment:
Ugh this is ugly. Nothing to really do about it now, but I've never been happy with the way we pass parameters to the dictionary builder... It would be nice to find the time to improve it, but it has never been a priority.

Reply:
Why is `notificationLevel ` a property of the training algorithm? Should it be a property of the process?

Reply:
Should we refactor all modules to use a shared g_displayLevel and routines? I started going in this direction, but @Cyan4973 made an argument that passing the display level around as function parameter makes it clear who needs to print. We should discuss and unify the approach as both arguments have merit.

Reply:
> Should we refactor all modules to use a shared g_displayLevel and routines?

This cannot be done for the code in `lib/`, which the dictBuilder resides in. The code in `lib/` must be thread-safe, so no global variables. Although it is a bit strange that the dictBuilder prints to output at all, none of our other library code does that.

I don't think that it makes sense to change the `notificationLevel` parameter on its own. A larger refactor is needed, and is beyond the scope of this PR. I think that the dictBuilder should ultimately not print to `stderr` at all, and should communicate its progress / status through other means, like a callback & better error codes.

Reply:
> I think that the dictBuilder should ultimately not print to `stderr` at all, and should communicate its progress / status through other means, like a callback & better error codes.

I fully agree.
`dictBuilder` is part of `lib`, it should be refactored to not depend on `<stdio>` (in `release` mode).
And yes, a callback is likely more appropriate to communicate progresses.
This is indeed a much more complex topic, not to be addressed here.